{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "961b239f-540b-492f-bfc6-19adfd712a1f",
   "metadata": {},
   "source": [
    "# Alphafold2\n",
    "\n",
    "## Introduction\n",
    "\n",
    "[AlphaFold](https://github.com/google-deepmind/alphafold) is a deep learning model developed by Google's AI company, DeepMind, to predict protein structures. It uses neural networks trained on known protein structures to estimate the 3D shape of proteins based on their amino acid sequences. AlphaFold made a significant breakthrough in 2020 by achieving unprecedented accuracy in the Critical Assessment of protein Structure Prediction (CASP) competition, effectively solving the long-standing protein folding problem. The model's ability to rapidly and accurately predict protein structures has wide-ranging implications for biological research, drug discovery, and understanding diseases at the molecular level.\n",
    "\n",
    "AlphaFold2's initial training focused on individual protein chains, making it highly proficient in predicting their structures. Subsequently, a variant called [AlphaFold-Multimer](https://doi.org/10.1101/2021.10.04.463034) was developed to address protein-protein complexes. This version can predict structures of both homo-multimers (complexes of identical protein chains) and hetero-multimers (complexes of different protein chains).\n",
    "\n",
    "Importantly, AlphaFold2 doesn't merely reproduce known protein structures. Independent studies have demonstrated its ability to predict previously unseen protein folds, expanding our understanding of protein structural diversity. This capability has been verified by multiple research groups, highlighting AlphaFold2's potential to uncover novel protein architectures not yet observed in the Protein Data Bank (PDB)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4c4f0f1-e30a-4476-8c27-6f51e36a3977",
   "metadata": {},
   "source": [
    "## Demo\n",
    "\n",
    "Below is a sample application, developed using [Gradio](https://www.gradio.app/), to facilitate the interaction with the [NVIDIA's Alphafold2 NIM](https://catalog.ngc.nvidia.com/orgs/nim/teams/deepmind/containers/alphafold2) container deployed to [OKE](https://www.oracle.com/cloud/cloud-native/kubernetes-engine/).\n",
    "\n",
    "### Prerequisites\n",
    "\n",
    "We need to install `gradio` Python package:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f8e305e-a948-4381-a5e5-d76f09919193",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install gradio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf8c8264-01d9-4eb2-925a-67b65e68f366",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import hashlib, json, requests, traceback\n",
    "from pathlib import Path\n",
    "\n",
    "import gradio as gr\n",
    "\n",
    "invoke_host = \"http://alphafold2:8000\"\n",
    "invoke_paths = {\n",
    "    \"Predict MSA from sequence\": \"/protein-structure/alphafold2/predict-msa-from-sequence\",\n",
    "    \"Predict structure from MSA\": \"/protein-structure/alphafold2/predict-structure-from-msa\",\n",
    "    \"Predict structure from sequence\": \"/protein-structure/alphafold2/predict-structure-from-sequence\",\n",
    "    # \"multimer_predict_msa_from_sequence\": \"/protein-structure/alphafold2/multimer/predict-msa-from-sequence\",\n",
    "    # \"multimer_predict_structure_from_msa\": \"/protein-structure/alphafold2/multimer/predict-structure-from-msa\",\n",
    "    # \"multimer_predict_structure_from_sequences\": \"/protein-structure/alphafold2/multimer/predict-structure-from-sequences\"\n",
    "}\n",
    "\n",
    "molstar_css = \"https://cdn.jsdelivr.net/npm/molstar@4.7.1/build/viewer/molstar.min.css\"\n",
    "molstar_js = \"https://cdn.jsdelivr.net/npm/molstar@4.7.1/build/viewer/molstar.min.js\"\n",
    "\n",
    "def update_view(output_selector, full_output):\n",
    "\n",
    "    try:\n",
    "        if output_selector and \"Structure\" in output_selector:\n",
    "            index = int(output_selector.split()[-1])\n",
    "        else:\n",
    "            return None\n",
    "            \n",
    "        structure = full_output[index-1]\n",
    "\n",
    "        js = f\"\"\"\n",
    "<head>\n",
    "    <link rel=\"stylesheet\" type=\"text/css\" href=\"{molstar_css}\" />\n",
    "</head>\n",
    "<body>\n",
    "    <div id=\"app\" style=\"position: relative;margin: 0 auto; height: 650px; width: 100%\"></div>\n",
    "    <script type=\"text/javascript\" src=\"{molstar_js}\"></script>\n",
    "    <script type=\"text/javascript\">\n",
    "    const pdbString = `\n",
    "{structure}\n",
    "`;        \n",
    "        molstar.Viewer.create(\"app\", {{\n",
    "            layoutIsExpanded: false,\n",
    "            layoutShowControls: false,\n",
    "            layoutShowRemoteState: false,\n",
    "            layoutShowSequence: true,\n",
    "            layoutShowLog: false,\n",
    "            layoutShowLeftPanel: false,\n",
    "            viewportShowExpand: false,\n",
    "            viewportShowSelectionMode: false,\n",
    "            viewportShowAnimation: false,\n",
    "        }}).then(viewer => {{\n",
    "            viewer.loadStructureFromData(pdbString, \"pdb\", false);\n",
    "        }});\n",
    "    </script>\n",
    "</body>\n",
    "\"\"\"\n",
    "\n",
    "        return f\"\"\"<iframe style=\"border:1;display:block;margin:auto;width: 100%; height: 700px\" srcdoc='{js}'></iframe>\"\"\"\n",
    "    except Exception as e:\n",
    "        raise gr.Error(\"Can't render the structure!\")\n",
    "        print(traceback.format_exc())\n",
    "    \n",
    "    return \"\"\n",
    "\n",
    "def predict_msa_from_sequence(\n",
    "    sequence,\n",
    "    algorithm,\n",
    "    bit_score,\n",
    "    databases,\n",
    "    e_value,\n",
    "    iterations):\n",
    "\n",
    "    try:\n",
    "        payload = {\n",
    "            \"sequence\": sequence,\n",
    "            \"algorithm\": algorithm,\n",
    "            \"bit_score\": bit_score,\n",
    "            \"databases\": databases,\n",
    "            \"e_value\": e_value,\n",
    "            \"iterations\": iterations\n",
    "        }\n",
    "        \n",
    "        headers = {\"Content-Type\": \"application/json\"}\n",
    "\n",
    "        response = requests.post(f\"{invoke_host}{invoke_paths['Predict MSA from sequence']}\", headers=headers, json=payload)\n",
    "        \n",
    "        if response.status_code == 200:\n",
    "            file_hash = hashlib.md5(json.dumps(response.json()).encode('utf-8')).hexdigest()\n",
    "            \n",
    "            alignments_file = Path(gr.utils.tempfile.gettempdir()).joinpath(f'{file_hash}-alignments.out').as_posix()\n",
    "            templates_file = Path(gr.utils.tempfile.gettempdir()).joinpath(f'{file_hash}-templates.out').as_posix()\n",
    "    \n",
    "            with open(alignments_file, 'wb') as f:\n",
    "                f.write(json.dumps(response.json().get('alignments')).encode('utf-8'))\n",
    "\n",
    "            with open(templates_file, 'wb') as f:\n",
    "                f.write(json.dumps(response.json().get('templates')).encode('utf-8'))\n",
    "                \n",
    "            return alignments_file, templates_file\n",
    "        elif response.status_code == 422:\n",
    "            raise gr.Warning(\"Invalid input!\")\n",
    "        else:\n",
    "            raise gr.Error(\"Could not process the request!\")\n",
    "   \n",
    "    except Exception:\n",
    "        print(traceback.format_exc())  \n",
    "        raise\n",
    "        \n",
    "    return None, None\n",
    "\n",
    "def predict_structure_from_msa(\n",
    "    sequence,\n",
    "    alignments,\n",
    "    templates,\n",
    "    relax_prediction):\n",
    "\n",
    "    try:\n",
    "        with open(alignments.get('path')) as alignments_file:\n",
    "            alignments_content = alignments_file.read()\n",
    "\n",
    "        with open(templates.get('path')) as templates_file:\n",
    "            templates_content = templates_file.read()\n",
    "            \n",
    "        payload = {\n",
    "            \"sequence\": sequence,\n",
    "            \"alignments\": json.loads(alignments_content),\n",
    "            \"templates\": json.loads(templates_content),\n",
    "            \"relax_prediction\": relax_prediction\n",
    "        }\n",
    "\n",
    "        headers = {\"Content-Type\": \"application/json\"}\n",
    "\n",
    "        response = requests.post(f\"{invoke_host}{invoke_paths['Predict structure from MSA']}\", headers=headers, json=payload)\n",
    "        \n",
    "        if response.status_code == 200:\n",
    "            file_hash = hashlib.md5(json.dumps(response.json()).encode('utf-8')).hexdigest()\n",
    "            \n",
    "            structure_file = Path(gr.utils.tempfile.gettempdir()).joinpath(f'{file_hash}-structure.out').as_posix()\n",
    "    \n",
    "            with open(structure_file, 'wb') as f:\n",
    "                f.write(json.dumps(response.json()).encode('utf-8'))\n",
    "            \n",
    "            returned_structures = [f'Structure {index+1}' for index, structure in enumerate(response.json())]\n",
    "            inference_output = response.json()\n",
    "            \n",
    "            return gr.Dropdown(choices=returned_structures, value=returned_structures[0]), inference_output, structure_file\n",
    "        \n",
    "        elif response.status_code == 422:\n",
    "            raise gr.Warning(\"Invalid input!\")\n",
    "        \n",
    "        else:\n",
    "            raise gr.Error(\"Could not process the request!\")\n",
    "        \n",
    "    except Exception:\n",
    "        print(traceback.format_exc())\n",
    "        raise\n",
    "        \n",
    "        \n",
    "    return None, None, None\n",
    "\n",
    "\n",
    "def predict_structure_from_sequence(\n",
    "    sequence,\n",
    "    algorithm,\n",
    "    bit_score,\n",
    "    databases,\n",
    "    e_value,\n",
    "    iterations,\n",
    "    relax_prediction):\n",
    "\n",
    "    try:\n",
    "        payload = {\n",
    "            \"sequence\": sequence,\n",
    "            \"algorithm\": algorithm,\n",
    "            \"bit_score\": bit_score,\n",
    "            \"databases\": databases,\n",
    "            \"e_value\": e_value,\n",
    "            \"iterations\": iterations,\n",
    "            \"relax_prediction\": relax_prediction\n",
    "        }\n",
    "\n",
    "        headers = {\"Content-Type\": \"application/json\"}\n",
    "\n",
    "        response = requests.post(f\"{invoke_host}{invoke_paths['Predict structure from sequence']}\", headers=headers, json=payload)\n",
    "        \n",
    "        if response.status_code == 200:\n",
    "            file_hash = hashlib.md5(json.dumps(response.json()).encode('utf-8')).hexdigest()\n",
    "            \n",
    "            structure_file = Path(gr.utils.tempfile.gettempdir()).joinpath(f'{file_hash}-structure.out').as_posix()\n",
    "    \n",
    "            with open(structure_file, 'wb') as f:\n",
    "                f.write(json.dumps(response.json()).encode('utf-8'))\n",
    "                \n",
    "            returned_structures = [f'Structure {index+1}' for index, structure in enumerate(response.json())]\n",
    "            inference_output = response.json()\n",
    "\n",
    "            return gr.Dropdown(choices=returned_structures, value=returned_structures[0]), inference_output, structure_file\n",
    "            \n",
    "        elif response.status_code == 422:\n",
    "            raise gr.Warning(\"Invalid input!\")\n",
    "        else:\n",
    "            raise gr.Error(\"Could not process the request!\")\n",
    "   \n",
    "    except Exception:\n",
    "        print(traceback.format_exc())  \n",
    "        raise\n",
    "        \n",
    "    return None, None, None\n",
    "\n",
    "with gr.Blocks(title=\"AlphaFold Sample App\", delete_cache=(172800, 172800)) as demo:\n",
    "    gr.Markdown(\"# AlphaFold Sample App\")\n",
    "\n",
    "    with gr.Tab(\"Predict MSA from sequence\", id=\"a\"):\n",
    "        gr.Markdown(\"## Perform a Multiple Sequence Alignment (MSA) and return the MSA and templates for alphafold inference.\")\n",
    "\n",
    "        t1_sequence = gr.Textbox(\n",
    "            label=\"Input amino acid sequence.\"\n",
    "        )\n",
    "\n",
    "        t1_databases = gr.Dropdown(\n",
    "            choices=[\"uniref90\", \"mgnify\", \"small_bfd\"],\n",
    "            value=[\"uniref90\", \"mgnify\", \"small_bfd\"],\n",
    "            multiselect=True,\n",
    "            interactive=True,\n",
    "            label=\"Select the databases to be used for Multiple Sequence Alignment prediction\",\n",
    "            info=\"Passing all three will provide the most accurate structural prediction at the cost of requiring the longest runtime.\"\n",
    "        )\n",
    "\n",
    "        t1_algorithm = gr.Dropdown(\n",
    "            choices=[\"jackhmmer\"],\n",
    "            value=\"jackhmmer\",\n",
    "            interactive=False,\n",
    "            label=\"The algorithm used for Multiple Sequence Alignment\"\n",
    "        )\n",
    "\n",
    "        t1_e_value = gr.Number(\n",
    "            value=0.0001,\n",
    "            label=\"The sequence e-value for filtering sequences in the MSA\",\n",
    "            info=\"Value should be between 0 and 1.\"\n",
    "        )\n",
    "\n",
    "        t1_bit_score = gr.Number(\n",
    "            value=0,\n",
    "            label=\"The sequence bit-score to use for filtering before MSA\",\n",
    "            info=\"If passed, this is used in place of e-value for filtering\"\n",
    "        )\n",
    "\n",
    "        t1_iterations = gr.Number(\n",
    "            value=1,\n",
    "            label=\"The number of MSA iterations to perform\"\n",
    "        )\n",
    "\n",
    "        gr.Examples(\n",
    "            [\n",
    "                [\n",
    "                    \"MAKTIKITQTRSAIGRLPKHKATLLGLGLRRIGHTVEREDTPAIRGMINAVSFMVKVEE\",\n",
    "                    [\"uniref90\", \"mgnify\", \"small_bfd\"],\n",
    "                    \"jackhmmer\",\n",
    "                    0.0001,\n",
    "                    0,\n",
    "                    1\n",
    "                ],\n",
    "            ],\n",
    "            [t1_sequence, t1_databases, t1_algorithm, t1_e_value, t1_bit_score, t1_iterations],\n",
    "        )\n",
    "\n",
    "        t1_run_button = gr.Button(\n",
    "            value=\"Run inference\")\n",
    "        \n",
    "        with gr.Accordion(visible=False) as t1_out_block:\n",
    "            t1_alignments_file = gr.File(label=\"Alignmnets Output File.\")\n",
    "            t1_templates_file = gr.File(label=\"Templates Output File.\")\n",
    "        \n",
    "        t1_run_button.click(lambda : gr.Accordion(visible=True), None, t1_out_block).then(\n",
    "            fn=predict_msa_from_sequence,\n",
    "            inputs=[t1_sequence, t1_algorithm, t1_bit_score, t1_databases, t1_e_value, t1_iterations], \n",
    "            outputs=[t1_alignments_file, t1_templates_file], \n",
    "            preprocess=False\n",
    "        )\n",
    "                    \n",
    "    with gr.Tab(\"Predict structure from MSA\", id=\"b\"):\n",
    "        gr.Markdown(\"## Perform structural prediction from an input MSA and templates.\")\n",
    "\n",
    "        t2_sequence = gr.Textbox(\n",
    "            label=\"Input amino acid sequence.\"\n",
    "        )\n",
    "\n",
    "        t2_alignments = gr.File(\n",
    "            label=\"The MSA results from predict-msa-from-sequence\",\n",
    "        )\n",
    "\n",
    "        t2_templates = gr.File(\n",
    "            label=\"Templates from the structural database search\",\n",
    "        )\n",
    "\n",
    "        t2_relax_prediction = gr.Checkbox(\n",
    "            label=\"Set to True to run structural relaxation after predictio\",\n",
    "            value=True\n",
    "        )\n",
    "        \n",
    "        gr.Examples(\n",
    "            [\n",
    "                [\n",
    "                    \"MAKTIKITQTRSAIGRLPKHKATLLGLGLRRIGHTVEREDTPAIRGMINAVSFMVKVEE\",\n",
    "                    \"examples/alignments.json\",\n",
    "                    \"examples/templates.json\",\n",
    "                    True\n",
    "                ],\n",
    "            ],\n",
    "            [t2_sequence, t2_alignments, t2_templates, t2_relax_prediction],\n",
    "        )\n",
    "\n",
    "        t2_run_button = gr.Button(\n",
    "            value=\"Run inference\")\n",
    "\n",
    "        t2_full_output = gr.State({})\n",
    "        \n",
    "        with gr.Accordion(visible=False) as t2_out_block:\n",
    "            t2_output_selector = gr.Dropdown(label=\"Returned Structures\", interactive=True)\n",
    "            t2_output_viewer = gr.HTML(label=\"Protein Structure Viewer\", show_label=True)\n",
    "            t2_output_file = gr.File(label=\"Inference result.\")\n",
    "        \n",
    "        t2_run_button.click(lambda : gr.Accordion(visible=True), None, t2_out_block).then(\n",
    "            fn=predict_structure_from_msa,\n",
    "            inputs=[t2_sequence, t2_alignments, t2_templates, t2_relax_prediction], \n",
    "            outputs=[t2_output_selector, t2_full_output, t2_output_file], \n",
    "            preprocess=False\n",
    "        )\n",
    "\n",
    "        t2_output_selector.change(fn=update_view,\n",
    "                inputs=[t2_output_selector, t2_full_output],\n",
    "                outputs=t2_output_viewer)\n",
    "         \n",
    "    with gr.Tab(\"Predict structure from sequence\", id=\"c\"):\n",
    "        gr.Markdown(\"## Predict a protein structure given an input amino acid sequence.\")\n",
    "\n",
    "        \n",
    "        t3_sequence = gr.Textbox(\n",
    "            label=\"Input amino acid sequence.\"\n",
    "        )\n",
    "\n",
    "        t3_databases = gr.Dropdown(\n",
    "            choices=[\"uniref90\", \"mgnify\", \"small_bfd\"],\n",
    "            value=[\"uniref90\", \"mgnify\", \"small_bfd\"],\n",
    "            multiselect=True,\n",
    "            interactive=True,\n",
    "            label=\"Select the databases to be used for Multiple Sequence Alignment prediction\",\n",
    "            info=\"Passing all three will provide the most accurate structural prediction at the cost of requiring the longest runtime.\"\n",
    "        )\n",
    "\n",
    "        t3_algorithm = gr.Dropdown(\n",
    "            choices=[\"jackhmmer\"],\n",
    "            value=\"jackhmmer\",\n",
    "            interactive=False,\n",
    "            label=\"The algorithm used for Multiple Sequence Alignment\"\n",
    "        )\n",
    "\n",
    "        t3_e_value = gr.Number(\n",
    "            value=0.0001,\n",
    "            label=\"The sequence e-value for filtering sequences in the MSA\",\n",
    "            info=\"Value should be between 0 and 1.\"\n",
    "        )\n",
    "\n",
    "        t3_bit_score = gr.Number(\n",
    "            value=0,\n",
    "            label=\"The sequence bit-score to use for filtering before MSA\",\n",
    "            info=\"If passed, this is used in place of e-value for filtering\"\n",
    "        )\n",
    "\n",
    "        t3_iterations = gr.Number(\n",
    "            value=1,\n",
    "            label=\"The number of MSA iterations to perform\"\n",
    "        )            \n",
    "\n",
    "        t3_relax_prediction = gr.Checkbox(\n",
    "            label=\"Relax prediction.\",\n",
    "            info=\"Set to True to run structural relaxation after prediction\",\n",
    "            value=True\n",
    "        )\n",
    "\n",
    "        gr.Examples(\n",
    "            [\n",
    "                [\n",
    "                    \"MAKTIKITQTRSAIGRLPKHKATLLGLGLRRIGHTVEREDTPAIRGMINAVSFMVKVEE\",\n",
    "                    [\"uniref90\", \"mgnify\", \"small_bfd\"],\n",
    "                    \"jackhmmer\",\n",
    "                    0.0001,\n",
    "                    0,\n",
    "                    1,\n",
    "                    True\n",
    "                ],\n",
    "            ],\n",
    "            [t3_sequence, t3_databases, t3_algorithm, t3_e_value, t3_bit_score, t3_iterations, t3_relax_prediction],\n",
    "        )\n",
    "        \n",
    "        t3_run_button = gr.Button(\n",
    "            value=\"Run inference\")\n",
    "\n",
    "        t3_full_output = gr.State({})\n",
    "        \n",
    "        with gr.Accordion(visible=False) as t3_out_block:\n",
    "            t3_output_selector = gr.Dropdown(label=\"Returned Structures\", interactive=True)\n",
    "            t3_output_viewer = gr.HTML(label=\"Protein Structure Viewer\", show_label=True)\n",
    "            t3_output_file = gr.File(label=\"Inference result.\")\n",
    "        \n",
    "        t3_run_button.click(lambda : gr.Accordion(visible=True), None, t3_out_block).then(\n",
    "            fn=predict_structure_from_sequence,\n",
    "            inputs=[t3_sequence, t3_algorithm, t3_bit_score, t3_databases, t3_e_value, t3_iterations, t3_relax_prediction], \n",
    "            outputs=[t3_output_selector, t3_full_output, t3_output_file], \n",
    "            preprocess=False\n",
    "        )\n",
    "\n",
    "        t3_output_selector.change(fn=update_view,\n",
    "                inputs=[t3_output_selector, t3_full_output],\n",
    "                outputs=t3_output_viewer)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    demo.launch(share=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}